documentation/smg/16-serialiser.md

```diff
@@ -0,0 +1,275 @@
+# Serialisation details
+
+As of the integration of the C++ serialiser, the serialised format has been updated to be more consistent and to have greater capabilities.
+
+For reference, the old format is also detailed below
+
+## New format
+
+### Standard format
+The new style has a standardised header format depending on the shape of the data (with the exception of function handles).
+This format allows extension to support arrays of objects and structs in a way which the old serialiser couldn't.
+It also greatly simplifies the tag structure and allows it to align with the C++ MEX API data types.
+Data in these cases is serialised in the linearised order as defined by `data(:)` and the array reshaped by the deserialiser.
+
+This standard format for the header is as follows:
+RANK |     tag    | dim1 | dim2 | ... | Data
+-----|------------|------|------|-----|------
+NULL | 32 + tag   | 0    |      |     |
+0    |    tag     |      |      |     | Data
+1    | 32 + tag   | nElem|      |     | Data
+2    | 64 + tag   | dim1 | dim2 |     | Data
+N    | N<<5 + tag | dim1 | dim2 | ... | Data
+
+Where `<<5` means bitshift left 5. (see [Tag Format][tag])
+
+- Tags are `uint8` data and the "magic" numbers are described in detail in the section [Tag Format][tag] below.
+- Dimensions are serialised as `uint32` 
+
+The single exception to this header is that of `function_handles` which are purely scalar and the rank component of the tag is used to specify the type of function handle:
+
+Handle type         |   tag   | value
+--------------------|---------|-------
+Simple/Class simple | 32 + 25 | 57
+Anonymous           | 64 + 25 | 89
+Scoped              | 96 + 25 | 121
+
+### Tag format
+
+The tag in the new format is an 1-byte tag where the top 3 bits give the rank of the data, the bottom 5 give the type-tag of the data
+
+| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
+|:---:|:--:|:--:|:--:|:-:|:-:|:-:|:-:|
+|  R  | R  | R  | T  | T | T | T | T |
+
+- R - Rank
+- T - Type
+
+### Serialisation formats in detail
+
+#### Simple data
+Simple data (tags 0-12) are serialised as:
```

Comment:
ie logical to uint64 in the table below. If the tag table comes above this, then reading this would be easier - also remembering that tag has mutiple meanings here for (a) type and (b) total 8-bit tag value (R+T)

It may be worth spelling out what the ranks NULL 0 and 1 actually mean by a couple of examples.

Reply:
I presume that rank of an array is an established and known term, I mean strictly we're not dealing with mathematical rank because we don't check irreducibility, but array rank is well known.

Reply:
The assumption is that rank is an array rank. It may have been a term in a different context. Saying array rank for the data could remove any ambiguity. Other than that, NULL and 1 appear to have the same tag, and the difference is in the size of the first dimension. If the rank is stored both in the first column (see comment above) and the tag, then the tag rank runs from 0 to 7 in options and I don't know what you do to represent NULL in the first column (what value?) and I don't know how you represent NULL in the 3 rank bits of the tag,

Reply:
If you added "type" to "tag" (so "type tags 0-12") then this would complete the reference to the header for the type tag table and to the T bits in the tag definition.